### PR TITLE
improvement(docker-backend): run Nemesis on docker backend in Jenkins

### PR DESCRIPTION
**Pull request by dimakr**
The change adds a capability to execute Nemesis cases on Docker backend in Jenkins.

Typically Docker backend is used for executing tests on an engineer's local environment. However, a standard test setup for a Scylla cluster - including the Scylla cluster itself, test runner, loader, monitoring instances - requires significant HW resources. This can sometimes prevent an engineer from being able to normally execute other tasks on the local machine, especially considering that a regular longevity+Nemesis test configuration runs for approximately 90 minutes.
To mitigate this problem, we can use the existing Jenkins pipelines to run a test runner instance for Docker backend in AWS cloud provider.

The following changes are introduced:
- longevity pipeline is updated to create only test runner instance when Docker is selected as a backend. This way the test runner instance is also serving the role of Docker backend node, the same way how tests are executed locally.
- configuration of storage disk is performed on creating an AWS test runner instance, when Docker is selected as a backend.
- sdcm.utils.nemesis.NemesisJobGenerator class is updated to be able to generate Jenkins job files for individual Nemesis cases;
- Individual Nemesis jenkins job files are generated for Docker backend.

The related task: https://github.com/scylladb/qa-tasks/issues/1636

### Testing
Test run in Jenkins on Docker backend: [CorruptThenScrubMonkey nemesis run on docker backend in Jenkins](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-docker-test/5/)
Test run in Jenkins on AWS (regression check): [CorruptThenScrubMonkey nemesis run on AWS in Jenkins](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-docker-test/4/)

### PR pre-checks (self review)
- [x] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevent to this change (if needed)

### CODE REVIEW

File: vars/createArgusTestRun.groovy [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
+    def cloud_provider = getCloudProviderFromBackend(params.backend)
     sh """#!/bin/bash
         set -xe
 
         echo "Creating Argus test run ..."
         if [[ -n "${params.requested_by_user}" ]] ; then
             export BUILD_USER_REQUESTED_BY=${params.requested_by_user}
         fi
-        export SCT_CLUSTER_BACKEND="${params.backend}"
+        export SCT_CLUSTER_BACKEND="${cloud_provider}"
```
Comment (fruch) [inline vars/createArgusTestRun.groovy:13]: Why are we changing this ? It doesn't seem to be related to this change set

File: sdcm/utils/nemesis.py [hunk 2]
```diff
@@ -113,8 +114,8 @@ def create_job_files_from_template(self) -> list[Path]:
             additional_configs = config.NEMESIS_REQUIRED_ADDITIONAL_CONFIGS.get(cls, [])
             additional_params = config.NEMESIS_ADDITIONAL_PIPELINE_PARAMS.get(cls, {})
             config_name = [
-                str(self.nemesis_test_config_dir / f"{self.base_job}-nemesis.yaml"),
```
Comment (fruch) [inline sdcm/utils/nemesis.py:116]: Why are you switching the order ?

File: sdcm/sct_runner.py [hunk 3]
```diff
@@ -600,8 +602,56 @@ def _create_instance(self,
...
+        if os.getenv("SCT_CLUSTER_BACKEND") == "docker":
+            LOGGER.info("Configure unused disks to use as a Docker container volume for Scylla data.")
+            LOGGER.info("Connecting to instance...")
+            remoter = self.get_remoter(host=instance.public_ip_address, connect_timeout=120)
+
+            LOGGER.info("Getting unused disks...")
+            disks = self._get_unused_disks(remoter)
+            LOGGER.info("The following unused disks are detected: %s", disks)
+            if disks:
+                self._setup_disks(remoter, disks)
```
Comment (fruch) [inline sdcm/sct_runner.py:615]: Missing the part we pointing docker to those mount points

File: configurations/nemesis/longevity-1gb-1h-nemesis.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
+test_duration: 90
```
Comment (fruch) [inline configurations/nemesis/longevity-1gb-1h-nemesis.yaml:1]: This file seems like a duplication,

And now we have 3 identical files with different names, we should align it and leave with only one.

**Review by fruch — COMMENTED** [review]
Looks in a good direction, few bits we need to address

File: vars/createArgusTestRun.groovy (hunk 1 above)
Comment (dimakr) [reply vars/createArgusTestRun.groovy:13]: This is my mistake.
Initial implementation for this PR used another word in Jenkins parameters, to indicate docker backend. And that word was not known in sdcm.SCTConfiguration.defaults_config_files map, so some logic was added to work around that.
But at the end simple 'docker' wording for indicating backend in Jenkins is working well for existing sct_config logic.
So this change is not needed, reverting it back.

File: configurations/nemesis/longevity-1gb-1h-nemesis.yaml (hunk 4 above)
Comment (dimakr) [reply configurations/nemesis/longevity-1gb-1h-nemesis.yaml:1]: Adding this file was done to reduce as much as possible the changes in NemesisJobGenerator logic.
The Job generator takes 'base-job' parameter and uses its name (e.g. 'longevity-1gb-1h') to find the base job, the base test config, etc. So we just add a new config and in the generator code we change nothing (or little), and then we are able to generate dedicated nemeses for a new backend.

Another option that was considered of how to adjust NemesisJobGenerator is to re-use 'configurations/nemesis/additional_configs/docker_backend_local.yaml' config. But it would require some more changes in JobGenerator.
Also it is less flexible in terms of parameters  - for instance at some point we can decide to use not i4i.xlarge instance type for SCTRunner on AWS, but 2xlarge, or 4xlarge, etc. This means we can also increase memory limits in 'append_scylla_args parameter', maybe increase test data volume from 1 GB to 5 GB, etc. Then it will not affect local docker backend runs experience if a separate config exists for Jenkins docker backend runs.

What do you think?

Comment (fruch) [reply configurations/nemesis/longevity-1gb-1h-nemesis.yaml:1]: I don't think we're gonna support multiple bases for the generator, if we'll need a certain setup we could build it, not to generate.


Regardless of the name wet should pick, we should have one copy of it

File: sdcm/utils/nemesis.py (hunk 2 above)
Comment (dimakr) [reply sdcm/utils/nemesis.py:116]: This is just to keep consistency in test resources naming.

Now all dedicated nemesis configs (the configurations/nemesis/*Monkey.yaml ones) have user prefix defined as 
e.g. 'user_prefix: 'longevity-5gb-1h-CorruptThenRepairMonkey''.
And if a user runs a docker backend on Jenkins, and sees that job uses configurations/nemesis/longevity-1gb-1h-nemesis.yaml base config, and in log sees that prepare stress command writes only 1G of data, but then sees that test resources are named as 'longevity-5gb-*-db-node-*' - it can cause confusion.

*Monkey.yaml configs are changed rarely (if ever). So when the 'longevity-1gb-1h-' (or similar) config is placed last in the list of configs for Job, we overwrite the user_prefix parameter value to exactly what the test does.

Comment (dimakr) [reply sdcm/utils/nemesis.py:116]: Reverting this as well. The custom user_prefix will come from docker_backend_local config, as we decided not to multiply base configs for Job generator.

File: sdcm/sct_runner.py (hunk 3 above)
Comment (dimakr) [reply sdcm/sct_runner.py]: Fixed this.
The docker now is on this additional disk:
```
ubuntu@ip-10-4-1-70:~/sct-results$ lsblk | grep mnt
nvme1n1      259:1    0 872.6G  0 disk /mnt/docker-persistent-storage

ubuntu@ip-10-4-1-70:~/sct-results$ docker inspect -f '{{ .Mounts }}' 15d4d598f324
[{bind  /var/run/docker.sock /var/run/docker.sock  rw true rprivate} {volume 93dd165165820f86c4ed24990106de312be3249acdaaaf27844ab17198c0db44 /mnt/docker-persistent-storage/volumes/93dd165165820f86c4ed24990106de312be3249acdaaaf27844ab17198c0db44/_data /var/lib/scylla local  true }]
```

Re-checked jobs:
Docker backend: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-docker-test/34/
AWS backend: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-docker-test/35/

File: vars/createArgusTestRun.groovy (hunk 1 above)
Comment (fruch) [reply vars/createArgusTestRun.groovy]: so please revert it

**Review by fruch — CHANGES_REQUESTED** [review]
LGTM, but need to revert this change https://github.com/scylladb/scylla-cluster-tests/pull/7322/commits/a508cdeab92450eed7f19ddeade7d79b8355481e#r1564750379

**Comment by soyacz** [comment]
> LGTM, but need to revert this change [a508cde#r1564750379](https://github.com/scylladb/scylla-cluster-tests/commit/a508cdeab92450eed7f19ddeade7d79b8355481e#r1564750379)

I think the `vars/createArgusTestRun.groovy` change was reverted, all the rest should be kept - unless you have something else in mind.

**Review by soyacz — DISMISSED** [review]
LGTM

File: sdcm/sct_runner.py [hunk 5]
```diff
@@ -600,8 +602,67 @@ def _create_instance(self,
         LOGGER.info("Got public IP: %s", instance.public_ip_address)
         self.instance = instance
 
+        # If the SCT runner instance is used as a Docker backend for tests
+        if os.getenv("SCT_CLUSTER_BACKEND") == "docker":
```
Comment (vponomaryov) [inline sdcm/sct_runner.py:606]: `self.params.get("cluster_backend")` must be used for it.
There is big enough probability to not have exactly env var be set, but yaml config option.

File: sdcm/sct_runner.py [hunk 6]
```diff
@@ -600,8 +602,67 @@ def _create_instance(self,
...
+            LOGGER.info("Configure unused disks to use as a persistent storage for Docker.")
+            LOGGER.info("Connecting to instance...")
+            remoter = self.get_remoter(host=instance.public_ip_address, connect_timeout=120)
+
+            LOGGER.info("Getting unused disks...")
```
Comment (vponomaryov) [inline sdcm/sct_runner.py:611]: Such kind of a message looks like a `LOGGER.debug(...)`, but not `info`.

File: sdcm/sct_runner.py [hunk 7]
```diff
@@ -600,8 +602,67 @@ def _create_instance(self,
         LOGGER.info("Got public IP: %s", instance.public_ip_address)
         self.instance = instance
 
+        # If the SCT runner instance is used as a Docker backend for tests
+        if os.getenv("SCT_CLUSTER_BACKEND") == "docker":
+            LOGGER.info("Configure unused disks to use as a persistent storage for Docker.")
+            LOGGER.info("Connecting to instance...")
```
Comment (vponomaryov) [inline sdcm/sct_runner.py:608]: Such kind of a message looks like a `LOGGER.debug(...)`, but not `info`.

File: sdcm/sct_runner.py [hunk 8]
```diff
@@ -600,8 +602,67 @@ def _create_instance(self,
...
+            remoter = self.get_remoter(host=instance.public_ip_address, connect_timeout=120)
+
+            LOGGER.info("Getting unused disks...")
+            disks = self._get_unused_disks(remoter)
+            LOGGER.info("The following unused disks are detected: %s", disks)
```
Comment (vponomaryov) [inline sdcm/sct_runner.py:613]: Same here

File: sdcm/sct_runner.py [hunk 9]
```diff
@@ -600,8 +602,67 @@ def _create_instance(self,
...
+
+        unused = []
+        for device in devices:
+            dev = device.replace("/dev/", '')
+            if len(glob.glob("/sys/class/block/{dev}/{dev}*".format(dev=dev))) == 0:
+                unused.append(device)
+        return unused
+
+    @staticmethod
+    def _setup_disks(remoter: Any, disks: list, mount_point: str) -> None:
```
Comment (vponomaryov) [inline sdcm/sct_runner.py:640]: Disk setup doesn't look as a requirement for being able to run a test on docker backend.
We do not expect perf to be great in this case.

Moreover, commit topic and description doesn't tell anything about the disk setup.
About such a significant change as changing disks formatting...

File: vars/createArgusTestRun.groovy [hunk 10]
```diff
@@ -2,15 +2,14 @@
...
-    def cloud_provider = getCloudProviderFromBackend(params.backend)
     sh """#!/bin/bash
         set -xe
 
         echo "Creating Argus test run ..."
         if [[ -n "${params.requested_by_user}" ]] ; then
             export BUILD_USER_REQUESTED_BY=${params.requested_by_user}
         fi
-        export SCT_CLUSTER_BACKEND="${cloud_provider}"
+        export SCT_CLUSTER_BACKEND="${params.backend}"
```
Comment (vponomaryov) [inline vars/createArgusTestRun.groovy:12]: This file change reverting must have been done in the proper commit.

File: sdcm/sct_runner.py (hunk 5 above)
Comment (dimakr) [reply sdcm/sct_runner.py:606]: The implemented logic is intended to be used only when Docker backend test run is started in Jenkins. In this case the env parameter is always set here https://github.com/scylladb/scylla-cluster-tests/blob/1f38a636ca1695c7bb0287dc9fd176dec7e192bf/vars/runSctTest.groovy#L28

And `self.params.get("cluster_backend")` won't help here. When running tests from Jenkins there is this differentiation between 'cluster backend for tests' and cloud provider for the 'cluster', like it is defined in https://github.com/scylladb/scylla-cluster-tests/blob/1f38a636ca1695c7bb0287dc9fd176dec7e192bf/vars/getCloudProviderFromBackend.groovy#L3

So, when the test is started from Jenkins, in the condition above we need to check what the selected backend for tests is (SCT_CLUSTER_BACKEND has this value), not the cloud provider (the value returned by `self.params.get("cluster_backend")`)

File: sdcm/sct_runner.py (hunk 9 above)
Comment (dimakr) [reply sdcm/sct_runner.py:640]: Without a dedicated physical disk Nemesis cases are always failing, when executed on Docker backend in Jenkins.
Only after configuring a free disk (which comes with e.g. an i4i instance type) to be used as a permanent storage for Docker - the Nemesis cases started to pass.

**Comment by dimakr** [comment]
> LGTM, but need to revert this change [a508cde#r1564750379](https://github.com/scylladb/scylla-cluster-tests/commit/a508cdeab92450eed7f19ddeade7d79b8355481e#r1564750379)

The change is reverted in the dependent commit.
Is it OK to have it this way, and we will have all the changes combined in the final squash?

File: sdcm/sct_runner.py (hunk 5 above)
Comment (vponomaryov) [reply sdcm/sct_runner.py:606]: @dimakr 
The `self.params.get("cluster_backend")` will always be equal to the `os.getenv("SCT_CLUSTER_BACKEND")` if the latter is set. And I proposed you to use the former approach with `self.params` to cover cases when env var is not set.
The `vars/getCloudProviderFromBackend.groovy` file is not related to the raised concern at all.

File: sdcm/sct_runner.py (hunk 9 above)
Comment (vponomaryov) [reply sdcm/sct_runner.py:640]: If nemesis that depends on the disk perf fails all the time without local NVMes then there should be detection of it and proper skipping.

**Comment by vponomaryov** [comment]
> > LGTM, but need to revert this change [a508cde#r1564750379](https://github.com/scylladb/scylla-cluster-tests/commit/a508cdeab92450eed7f19ddeade7d79b8355481e#r1564750379)
> 
> @fruch The change is reverted in the dependent commit. Is it OK to have it this way, and we will have all the changes combined in the final squash?

We use `rebase and merge` and not `squash and merge` actions.
We don't want to lose git commit history details.

Then, in your case it is pretty simple action - need to revert whole file changes.

**Comment by fruch** [comment]
> > > LGTM, but need to revert this change [a508cde#r1564750379](https://github.com/scylladb/scylla-cluster-tests/commit/a508cdeab92450eed7f19ddeade7d79b8355481e#r1564750379)
> > 
> > 
> > @fruch The change is reverted in the dependent commit. Is it OK to have it this way, and we will have all the changes combined in the final squash?
> 
> We use `rebase and merge` and not `squash and merge` actions. We don't want to lose git commit history details.
> 
> Then, in your case it is pretty simple action - need to revert whole file changes.

@dimakr 

It might sound a bit pedantic, but yes we are striving to have a clear commits flow, especially since we might need to backport fixes/features into release branches 

Also there are some more comments to address about a usage of environmental variable, and some logs level

In less you have a very strong argument, complying to what maintener asks is usually the quickest way (they even guard me, from cutting some corners)

File: sdcm/sct_runner.py (hunk 5 above)
Comment (dimakr) [reply sdcm/sct_runner.py]: Changes to use params for getting cluster_backend value.

File: sdcm/sct_runner.py (hunk 7 above)
Comment (dimakr) [reply sdcm/sct_runner.py]: Changed to debug instead of the very first message where we notify about 'configuring'.
It helps to see immediately in that this step was executed, e.g. in Jenkins, without the need to download and look through test logs

File: sdcm/sct_runner.py (hunk 9 above)
Comment (dimakr) [reply sdcm/sct_runner.py:639]: Updated the commit message.
With regard to the configuring the dedicated disk - no Nemesis case is able to run on EBS, So the dedicated disk is necessary to be able to run docker backend on AWS.

File: vars/createArgusTestRun.groovy (hunk 10 above)
Comment (dimakr) [reply vars/createArgusTestRun.groovy]: Reverted in the proper commit.

**Comment by dimakr** [comment]
The comments are addressed.
This is the sanity checks that the latest changes do not brake the functionality: [longevity-5gb-1h-docker-test_build_48](https://jenkins.scylladb.com/job/scylla-staging/job/dimakr/job/longevity-5gb-1h-docker-test/48/)

**Review by soyacz — APPROVED** [review]
LGTM

**Review by fruch — APPROVED** [review]
LGTM
